### DISCUSSION TITLE
Black-Litterman with factor risk but views on securities

### DISCUSSION THREAD

**Discussion by nemoincognito** [skfolio/skfolio#54] [Ideas]
Lets say you had a view on securities in the classic Black-Litterman format: NVDA vs AMD, outright views on something else etc for your mu values. All your securities have some factor exposures which you can estimate using the FactorModel class quite well. How would you do Black-Litterman on the factors but also integrate the views at the security level? Is this possible?

**Comment by HugoDelatte** [comment]
For the below example I had to push a fix in version 0.2.1 so to run it you need to upgrade to the latest version.

Here is a solution:

```python

from skfolio.datasets import load_factors_dataset, load_sp500_dataset
from skfolio.preprocessing import prices_to_returns
from skfolio.prior import BlackLitterman, FactorModel

prices = load_sp500_dataset()
factor_prices = load_factors_dataset()
X, y = prices_to_returns(prices, factor_prices)

factor_views = [
    "SIZE == 0.00039",
    "SIZE - VLUE == 0.00011 ",
    "MTUM - QUAL == 0.00007",
]

assets_views = [
    "AAPL == 0.00098",
    "AAPL - GE == 0.00086",
    "JPM - GE == 0.00059",
]

model = BlackLitterman(
    views=assets_views,
    prior_estimator=FactorModel(
        factor_prior_estimator=BlackLitterman(views=factor_views),
    )
)

model.fit(X, y)

print(model.prior_model_.mu)
print(model.prior_model_.covariance)
```

Here is what's happening for the expected returns (same logic for the covariance):

1. get the empirical factor's mu of shape (5,)
2. apply BL with views on the factors to get the BL posterior factor's mu of shape (5,)
3. apply a factor model to get the loading matrix of shape (20, 5) (betas for each assets) and get the assets' mu of shape (20,) by multiply the loading matrix by the factor's mu
4. apply a BL with views on the assets to get the BL posterior asset's mu of shape (20,)

As you can see, BL is applied twice because we have the factor model step in the middle. I'm not aware of a paper or implementation that do it in one single step but if you are please let me know and we will have a look.
